### PR TITLE
Swift 4 Update

### PR DESCRIPTION
**Pull request by danieleftodi**
Hi, @stuffrabbit,

I updated the code to Swift 4.2.

I have also added an Xcode Workspace to the project. This will make it easier to work, and add other Targets too.

I'm planing on making:
1) Convert this code into a proper Swift Package (to make it easier for everyone to add it to their projects...)
2) macOS Command Line version
3) a Vapor Server Side Swift

### CODE REVIEW

File: Example_Projects/SwiftCoAPClientExample/SwiftCoAPClientExample/SCMessage.swift [hunk 1]
```diff
@@ -34,7 +34,7 @@ protocol SCCoAPTransportLayerDelegate: class {
 
 protocol SCCoAPTransportLayerProtocol: class {
     //SCClient uses this property to assign itself as delegate
-    weak var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
+    var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
```
Comment (stuffrabbit) [inline Example_Projects/SwiftCoAPClientExample/SwiftCoAPClientExample/SCMessage.swift:37]: why is `weak` gone here? I guess we still need it.

File: Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCMessage.swift [hunk 2]
```diff
@@ -34,7 +34,7 @@ protocol SCCoAPTransportLayerDelegate: class {
 
 protocol SCCoAPTransportLayerProtocol: class {
     //SCClient uses this property to assign itself as delegate
-    weak var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
+    var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
```
Comment (stuffrabbit) [inline Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCMessage.swift:37]: Same as above?

File: Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift [hunk 3]
```diff
@@ -57,7 +57,7 @@ class SCServer: NSObject {
     fileprivate class SCAddressWrapper: NSObject {
         let hostname: String
         let port: UInt16
-        override var hashValue: Int {
+        override var hash: Int {
```
Comment (stuffrabbit) [inline Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift:60]: You don't like hashValue? :)

File: Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift [hunk 4]
```diff
@@ -429,7 +429,8 @@ class SCServer: NSObject {
                         valueArray[index] = (message.token, hostname, port, newSequenceNumber, prefferredBlock2SZX)
                     }
                     else {
-                        valueArray += [(message.token, hostname, port, 0, prefferredBlock2SZX)]
+//                        valueArray += [(message.token, hostname, port, 0, prefferredBlock2SZX)]
```
Comment (stuffrabbit) [inline Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift]: please remove this comment.

File: SwiftCoAP_Library/SCMessage.swift [hunk 5]
```diff
@@ -34,7 +34,7 @@ protocol SCCoAPTransportLayerDelegate: class {
 
 protocol SCCoAPTransportLayerProtocol: class {
     //SCClient uses this property to assign itself as delegate
-    weak var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
+    var transportLayerDelegate: SCCoAPTransportLayerDelegate! { get set }
```
Comment (stuffrabbit) [inline SwiftCoAP_Library/SCMessage.swift:37]: Same as above

File: SwiftCoAP_Library/SCServer.swift [hunk 6]
```diff
@@ -57,7 +57,7 @@ class SCServer: NSObject {
     fileprivate class SCAddressWrapper: NSObject {
         let hostname: String
         let port: UInt16
-        override var hashValue: Int {
+        override var hash: Int {
```
Comment (stuffrabbit) [inline SwiftCoAP_Library/SCServer.swift:60]: same as above

File: SwiftCoAP_Library/SCServer.swift [hunk 7]
```diff
@@ -429,7 +429,8 @@ class SCServer: NSObject {
                         valueArray[index] = (message.token, hostname, port, newSequenceNumber, prefferredBlock2SZX)
                     }
                     else {
-                        valueArray += [(message.token, hostname, port, 0, prefferredBlock2SZX)]
+//                        valueArray += [(message.token, hostname, port, 0, prefferredBlock2SZX)]
```
Comment (stuffrabbit) [inline SwiftCoAP_Library/SCServer.swift]: Comment can also go away here.

**Review by stuffrabbit — CHANGES_REQUESTED** [review]
Thanks for your contribution! Can you just quickly explain why exactly we need to add the workspace?

File: Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift (hunk 3 above)
Comment (danieleftodi) [reply Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift:60]: Xcodee Warning: 
```Override of 'NSObject.hashValue' is deprecated; override 'NSObject.hash' to get consistent hashing behavior```

hashValue changed from swift 3 --> hash in swift 4 ...

File: SwiftCoAP_Library/SCMessage.swift (hunk 5 above)
Comment (danieleftodi) [reply SwiftCoAP_Library/SCMessage.swift:37]: Xcode Warning:
`'weak' should not be applied to a property declaration in a protocol and will be disallowed in future versions`

File: Example_Projects/SwiftCoAPClientExample/SwiftCoAPClientExample/SCMessage.swift (hunk 1 above)
Comment (danieleftodi) [reply Example_Projects/SwiftCoAPClientExample/SwiftCoAPClientExample/SCMessage.swift:37]: Xcode Warning:
`'weak' should not be applied to a property declaration in a protocol and will be disallowed in future versions`

File: Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCMessage.swift (hunk 2 above)
Comment (danieleftodi) [reply Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCMessage.swift:37]: Xcode Warning:
`'weak' should not be applied to a property declaration in a protocol and will be disallowed in future versions`

File: Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift (hunk 4 above)
Comment (danieleftodi) [reply Example_Projects/SwiftCoAPServerExample/SwiftCoAPServerExample/SCServer.swift]: Done in next commit.

File: SwiftCoAP_Library/SCServer.swift (hunk 6 above)
Comment (danieleftodi) [reply SwiftCoAP_Library/SCServer.swift:60]: Xcodee Warning: 
```Override of 'NSObject.hashValue' is deprecated; override 'NSObject.hash' to get consistent hashing behavior```

hashValue changed from swift 3 --> hash in swift 4 ...

File: SwiftCoAP_Library/SCServer.swift (hunk 7 above)
Comment (danieleftodi) [reply SwiftCoAP_Library/SCServer.swift]: Done in next commit.

**Comment by danieleftodi** [comment]
> Can you just quickly explain why exactly we need to add the workspace?

As you can see from the attached screenshot of the workspace. It makes it easier to work with two sample projects, and maintaining the "Library" ... all three at the same time. 

Also, keep in mind, I'm working on converting "SwiftCoAP_Library"-folder into a "SwiftPM" (Swift Package) 😊 ... so working with multiple projects and a "SwiftPM" in one "Workspace" make it a bit easier to jump around in the entire project ...

Instead of having to switch between multiple windows / applications (Remember that the "SwiftCoAP_Library" and Readme & License... wasn't part of any project file ... so one had to open those files in ... say Sublime? ... to work with those files ... 

All in all, if you ask me, IMHO ... the moment one starts using more then one Xcode project and working with "some sort of package library" structure ... one should always strive to gather the entire project into a workspace 🤓

**Comment by stuffrabbit** [comment]
Thanks for your contribution!

**Pull request merged**